src/engraving/dom/laissezvib.cpp

```diff
@@ -0,0 +1,283 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-Studio-CLA-applies
+ *
+ * MuseScore Studio
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore Limited
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "laissezvib.h"
+#include "chord.h"
+#include "dom/measure.h"
+#include "dom/mscoreview.h"
+#include "dom/score.h"
+#include "note.h"
+#include "staff.h"
+#include "style/style.h"
+
+using namespace mu::engraving;
+
+static const ElementStyle laissezVibStyle {
+    { Sid::minLaissezVibLength, Pid::MIN_LENGTH }
+};
+
+LaissezVib::LaissezVib(Note* parent)
+    : Tie(ElementType::LAISSEZ_VIB, parent)
+{
+    initElementStyle(&laissezVibStyle);
+}
+
+mu::engraving::PropertyValue mu::engraving::LaissezVib::getProperty(Pid propertyId) const
+{
+    switch (propertyId) {
+    case Pid::MIN_LENGTH:
+        return minLength();
+    default:
+        return Tie::getProperty(propertyId);
+    }
+}
+
+PropertyValue LaissezVib::propertyDefault(Pid propertyId) const
+{
+    switch (propertyId) {
+    case Pid::MIN_LENGTH:
+        return Spatium(2.0);
+    default:
+        return Tie::propertyDefault(propertyId);
+    }
+}
+
+bool LaissezVib::setProperty(Pid propertyId, const PropertyValue& v)
+{
+    switch (propertyId) {
+    case Pid::MIN_LENGTH:
+        setMinLength(v.value<Spatium>());
+        break;
+    default:
+        return Tie::setProperty(propertyId, v);
+    }
+    triggerLayout();
+    return true;
+}
+
+SymId LaissezVib::symId() const
+{
+    return up() ? SymId::articLaissezVibrerAbove : SymId::articLaissezVibrerBelow;
+}
+
+static int compareNotesPos(const Note* n1, const Note* n2)
+{
+    if (n1->line() != n2->line() && !(n1->staffType()->isTabStaff())) {
+        return n2->line() - n1->line();
+    } else if (n1->string() != n2->string()) {
+        return n2->string() - n1->string();
+    } else {
+        return n1->pitch() - n2->pitch();
+    }
+}
+
+void LaissezVib::calculateDirection()
```

Comment:
Couple of thoughts (referred to all three of these functions, namely calculateDirection, calculateIsInside, isOuterTieOfChord).
1. I wonder if it would be worth trying to factor out some of the reused code here. I can see that the logic is quite similar to normal ties, with the obvious difference that we aren't looking at the end chord. The answer may well be no: sometimes avoiding repetitions _at all costs_ can result in messier code. But do give it a thought, if you can see any easy way of doing it. A pattern that I've found useful sometimes when dealing with overrides like this is to separate the part that's common from the part that's overridden (see for instance Spanner::computeEndElement and Spanner::doComputeEndElement to see what I mean).
2. Independently from the previous point: perhaps calculateDirection, which is the more strictly layout-dependend one, is worth moving out of the DOM and into the layout module?

Reply:
This and the point above became immediately obvious when beginning partial ties!  I'll tidy these up and just have single methods in `Tie` which handle both start & end or only one or the other.